cf-net/cf-net.c

```diff
@@ -0,0 +1,766 @@
+            (uintmax_t) sb.st_mode,  (intmax_t)  sb.st_size,
+            (uintmax_t) sb.st_uid,   (uintmax_t) sb.st_gid,
+            (intmax_t)  sb.st_atime, (intmax_t)  sb.st_mtime);
+        CFNetStatPrint(file, sb.st_mode, hostname);
+    }
+
+    DisconnectServer(conn);
+    return 0;
+}
+
+static int invalid_command(const char *cmd)
+{
+    Log(LOG_LEVEL_ERR, "Invalid, see: 'cf-net help %s' for more info\n", cmd);
+    return -1;
+}
+
+
+static int CFNetGet(CFNetOptions *opts, const char* hostname, char **args)
+{
+    assert(opts);
+    assert(hostname);
+    assert(args);
+    AgentConnection *conn = CFNetOpenConnection(hostname);
+    if (conn == NULL)
+    {
+        return 1;
+    }
+    char *local_file = NULL;
+
+    // TODO: Propagate argv and argc from main()
+    int argc = 0;
+    while (args[argc] != NULL)
+    {
+        ++argc;
+    }
+
+    static struct option longopts[] = {
+         { "output",     required_argument,      NULL,           'o' },
+         { NULL,         0,                      NULL,           0   }
+    };
+    assert(opts != NULL);
+    if (argc <= 1)
+    {
+        return invalid_command("get");
+    }
+    extern int optind;
+    optind = 0;
+    extern char *optarg;
+    int c = 0;
+    int start_index = 1;
+    // TODO: Experiment with more user friendly leading - optstring
+    const char *optstr = "+o:"; // + means stop for non opt arg. :)
+    bool specified_path = false;
+    while ((c = getopt_long(argc, args, optstr, longopts, &start_index))
+            != -1)
+    {
+        switch (c)
+        {
+            case 'o':
+            {
+                local_file = xstrdup(optarg);
+                specified_path = true;
+                break;
+            }
+            case ':':
+            {
+                return invalid_command("get");
+                break;
+            }
+            case '?':
+            {
+                return invalid_command("get");
+                break;
+            }
+            default:
+            {
+                printf("Default optarg = '%s', c = '%c' = %i",
+                       optarg, c, (int)c);
+                break;
+            }
+        }
+    }
+    args = &(args[optind]);
+    argc -= optind;
+    const char *remote_file = args[0];
+
+    if(local_file == NULL)
+    {
+        char *last = strrchr(remote_file, '/');
+        if (last != NULL)
+        {
+            local_file = xstrdup(last + 1);
+        }
+        else
+        {
+            local_file = xstrdup(remote_file);
+        }
+    }
+
+    if (specified_path && strcmp(local_file, "-") == 0)
+    {
+         // TODO: Should rewrite CopyRegularFileNet etc. to take fd argument
+         // and a simple helper function to open a file as well.
+        printf("Output to stdout not yet implemented (TODO)");
+        free(local_file);
+        return -1;
+    }
+
+    struct stat sb;
+    int r = cf_remote_stat(conn, true, remote_file, &sb, "file");
+    if (r!= 0)
```

Comment:
Turns out there are two! You fixed the one in `CFNetStat`, but this one lingers.